keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers/README.md

```diff
@@ -347,6 +349,30 @@ correct order.
 [enables]:https://github.com/kubernetes/kubernetes/blob/220498b83af8b5cbf8c1c1a012b64c956d3ebf9b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L703
 [filter]:#updating-storageversion
 
+#### Agreements
+
+1. Storageversion updates will be triggered by CRD creates and updates
+2. We will perform async updates of storageversions such that we do not block writes of other unrelated APIs(for ex, crdA writes waiting for update of crdB's storageversion). These requests should be handled asynchronously
+3. We will block a storageversion update until the teardown of prev storage is finished
+   1. this means we will wait to update storageversion of a CRD until
+      1. old storage of the same CRD is deleted
+      2. pending CR writes using old storageversion of the same CRD are completed
+   2. We will do this to prevent publishing a newer storageversion while a pending CR write is still in progress. Otherwise, if the pending CR write finishes writing the CR using an old version, the storageversion API would not reflect that and the object will remain in etcd, encoded in an old version forever 
+4. We will block new CR writes for a CRD until we have published its latest storageversion. This is discussed more in the [limitations] section
+
+[limitations]:#Limitations 
+
+#### Limitations
+
+When a storageversion of a CRD is updated, we will ensure that all new CR writes:
+1. wait for the latest storageversion to be published
```

Comment:
If a CRD is being updated to store v2 instead of v1, and two requests come in at the same time:
1. Create a new CR
2. Update the CRD to store v3 instead of v2

Which one is the latest storageversion here? v2 or v3?

Reply:
For "at the same time": without a happens-before constraint, I don't think we should define the behavior.

Our guarantee should be that once a response to the CRD Update is sent, a happens-before is established, and all future CR creates store v3 (because clients can observe this ordering).  After the CRD update is received and before the response is sent,  storage at either version is possible.

Reply:
Depends on which request will be processed by the apiserver first. Following scenarios are possible:
1. CR request is received first: it will wait for the v2 update to finish. 
        
    1a. If no new CRD update is received yet, storageversion is successfully updated to v2, and the CR is written in v2. Now the CRD update to v3 is received, and the storageversion is updated to v3.
        
    1b. If a new CRD update to v3 is received at this time, our async SV update loop will switch to processing the latest storageversion - v3, and will never process the v2 update. CR writes waiting for v2 update to finish will time out.

2. CRD request is received first: our async SV update loop switches to processing the v3 update. IF now a CR request is received, it will wait for v3 update to finish and will be served in v3.